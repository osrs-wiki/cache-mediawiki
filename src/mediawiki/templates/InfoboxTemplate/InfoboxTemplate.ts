import {
  MediaWikiTemplate,
  MediaWikiContent,
} from "@osrs-wiki/mediawiki-builder";

import { Template } from "./InfoboxTemplate.types";

import { capitalize } from "@/utils/strings";

export class InfoboxTemplate<T> extends Template {
  params: T | T[];

  constructor(name: string, params: T | T[]) {
    super(`Infobox ${capitalize(name)}`);
    this.params = params;
  }

  build() {
    const infoboxTemplate = new MediaWikiTemplate(this.name);

    if (Array.isArray(this.params)) {
      this.buildMultipleParams(infoboxTemplate, this.params);
    } else {
      this.buildSingleParams(infoboxTemplate, this.params);
    }

    return infoboxTemplate;
  }

  private buildSingleParams(template: MediaWikiTemplate, params: T) {
    Object.keys(params).forEach((key) => {
      const value = params[key as keyof T];
      if (value !== undefined) {
        const parsedValue = this.parseValue(value);
        template.add(key, parsedValue);
      }
    });
  }

  private buildMultipleParams(template: MediaWikiTemplate, paramsArray: T[]) {
    if (paramsArray.length === 0) return;

    // Add version parameters first
    paramsArray.forEach((_, index) => {
      template.add(`version${index + 1}`, (index + 1).toString());
    });

    // Get all unique keys across all param sets
    const allKeys = new Set<string>();
    paramsArray.forEach((params) => {
      Object.keys(params).forEach((key) => allKeys.add(key));
    });

    // For each key, determine if values are shared or different
    allKeys.forEach((key) => {
      const values = paramsArray.map((params) => {
        const value = params[key as keyof T];
        return value !== undefined ? this.parseValue(value) : undefined;
      });

      // Filter out undefined values for comparison
      const definedValues = values.filter((v) => v !== undefined);

      if (definedValues.length === 0) return;

      // Check if all defined values are the same
      const firstValue = definedValues[0];
      const allSame = definedValues.every((v) => v === firstValue);

      if (allSame) {
        // All defined values are the same - use shared parameter
        template.add(key, firstValue as string);
      } else {
        // Values differ - use numbered parameters
        values.forEach((value, index) => {
          if (value !== undefined) {
            template.add(`${key}${index + 1}`, value);
          }
        });
      }
    });
  }

  private parseValue(value: unknown): string {
    if (typeof value === "boolean") {
      return value ? "Yes" : "No";
    } else if (value instanceof MediaWikiContent) {
      return value.build();
    } else if (Array.isArray(value)) {
      if (value.every((v) => typeof v === "string" || typeof v === "number")) {
        return value.filter((v) => v && v !== null).join(", ");
      } else {
        return value
          .filter((v) => v && v !== null)
          .map((v) => (v instanceof MediaWikiContent ? v.build() : `${v}`))
          .join(" ");
      }
    } else if (value !== undefined && value !== null) {
      return `${value}`;
    }
    return "";
  }
}
