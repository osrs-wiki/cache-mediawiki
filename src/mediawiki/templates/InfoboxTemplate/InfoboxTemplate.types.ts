export { InfoboxTemplateOptions, Template } from "../templates.types";