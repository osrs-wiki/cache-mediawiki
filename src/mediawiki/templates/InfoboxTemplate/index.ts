export { InfoboxTemplate } from "./InfoboxTemplate";
export * from "./InfoboxTemplate.types";