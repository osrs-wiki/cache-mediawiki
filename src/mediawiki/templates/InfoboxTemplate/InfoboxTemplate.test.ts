import { InfoboxTemplate } from "./InfoboxTemplate";

describe("InfoboxTemplate", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe("single params", () => {
    it("should build template with single params", () => {
      const template = new InfoboxTemplate("test", {
        name: "Test Name",
        level: 10,
        active: true,
        disabled: false,
      });

      const result = template.build();
      expect(result.name).toBe("Infobox Test");
      expect(result.build()).toMatchSnapshot();
    });
  });

  describe("multiple params", () => {
    it("should build template with shared parameters when values are the same", () => {
      const template = new InfoboxTemplate("monster", [
        { name: "Guard", level: 21, location: "Varrock" },
        { name: "Guard", level: 21, location: "Varrock" },
      ]);

      const result = template.build();
      expect(result.build()).toMatchSnapshot();
    });

    it("should build template with numbered parameters when values differ", () => {
      const template = new InfoboxTemplate("monster", [
        { name: "Guard", level: 21, examine: "He tries to keep order." },
        { name: "Guard", level: 21, examine: "She tries to keep order." },
      ]);

      const result = template.build();
      expect(result.build()).toMatchSnapshot();
    });

    it("should handle mixed shared and numbered parameters", () => {
      const template = new InfoboxTemplate("monster", [
        { name: "Guard", level: 21, examine: "He tries to keep order.", id: "1001" },
        { name: "Guard", level: 22, examine: "She tries to keep order.", id: "1002" },
      ]);

      const result = template.build();
      expect(result.build()).toMatchSnapshot();
    });

    it("should handle undefined values properly", () => {
      const template = new InfoboxTemplate("monster", [
        { name: "Guard", level: 21, examine: "He tries to keep order." },
        { name: "Guard", level: 21 }, // missing examine
      ]);

      const result = template.build();
      expect(result.build()).toMatchSnapshot();
    });
  });
});