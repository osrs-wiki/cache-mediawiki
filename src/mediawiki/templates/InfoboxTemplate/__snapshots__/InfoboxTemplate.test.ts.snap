// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfoboxTemplate multiple params should build template with numbered parameters when values differ 1`] = `
"{{Infobox Monster
|version1 = 1
|version2 = 2
|name = Guard
|level = 21
|examine1 = He tries to keep order.
|examine2 = She tries to keep order.
}}
"
`;

exports[`InfoboxTemplate multiple params should build template with shared parameters when values are the same 1`] = `
"{{Infobox Monster
|version1 = 1
|version2 = 2
|name = Guard
|level = 21
|location = Varrock
}}
"
`;

exports[`InfoboxTemplate multiple params should handle mixed shared and numbered parameters 1`] = `
"{{Infobox Monster
|version1 = 1
|version2 = 2
|name = Guard
|level1 = 21
|level2 = 22
|examine1 = He tries to keep order.
|examine2 = She tries to keep order.
|id1 = 1001
|id2 = 1002
}}
"
`;

exports[`InfoboxTemplate multiple params should handle undefined values properly 1`] = `
"{{Infobox Monster
|version1 = 1
|version2 = 2
|name = Guard
|level = 21
|examine1 = He tries to keep order.
}}
"
`;

exports[`InfoboxTemplate single params should build template with single params 1`] = `
"{{Infobox Test
|name = Test Name
|level = 10
|active = Yes
|disabled = No
}}
"
`;
